node/subsystem-util/src/lib.rs

```diff
@@ -209,6 +210,53 @@ specialize_requests! {
 	fn request_validation_code_hash(para_id: ParaId, assumption: OccupiedCoreAssumption)
 		-> Option<ValidationCodeHash>; ValidationCodeHash;
 	fn request_on_chain_votes() -> Option<ScrapedOnChainVotes>; FetchOnChainVotes;
+	fn request_session_executor_params(session_index: SessionIndex) -> Option<ExecutorParams>; SessionExecutorParams;
+}
+
+/// Requests executor parameters from the runtime effective at given relay-parent. First obtains
+/// session index at the relay-parent, relying on the fact that it should be cached by the runtime
+/// API caching layer even if the block itself has already been pruned. Then requests executor
+/// parameters by session index.
+/// Returns an error if failed to communicate to the runtime, or the parameters are not in the
+/// storage, which should never happen.
+/// Returns default execution parameters if the runtime doesn't yet support `SessionExecutorParams`
+/// API call.
+/// Otherwise, returns execution parameters returned by the runtime.
+pub async fn executor_params_at_relay_parent(
+	relay_parent: Hash,
+	sender: &mut impl overseer::SubsystemSender<RuntimeApiMessage>,
+) -> Result<ExecutorParams, Error> {
+	match request_session_index_for_child(relay_parent, sender).await.await {
+		Err(err) => {
+			// Failed to communicate with the runtime
+			Err(Error::Oneshot(err))
+		},
+		Ok(Err(err)) => {
+			// Runtime has failed to obtain a session index at the relay-parent; should never happen
+			Err(Error::RuntimeApi(err))
+		},
+		Ok(Ok(session_index)) => {
+			match request_session_executor_params(relay_parent, session_index, sender).await.await {
+				Err(err) => {
+					// Failed to communicate with the runtime
+					Err(Error::Oneshot(err))
+				},
+				Ok(Err(_)) => {
```

Comment:
Any error means not supported yet? Shouldn't we check the version to determine that? We should only return default parameters if really not _yet_ supported. Returning default parameters on any other error could cause consensus issues.

Reply:
As far as I can see `specialize_requests` generates code which queries the runtime-api via the runtime-api-subsystem. The version check is performed there and the subsystem returns `Err(RuntimeApiError::NotSupported` . This is [here](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/core/runtime-api/src/lib.rs#L381).

The match here should explicitly check for this error value and return default in this case, as you suggest.

Reply:
Fixed.

Reply:
I'm bringing back this discussion because of today's incident on Versi when trying to burn in the PR, which resulted in the network being practically brought down by a storm of disputes.

The behavior pointed out by @tdimitrov only applies when the runtime does not support requesting the API version. When a real runtime method is not supported, another error is returned: `RuntimeApiError::Execution { runtime_api_name: "session_executor_params", source: Application(Execution(Other("Exported method ParachainHost_session_executor_params is not found"))) }`.

That's due to the Substrate executor logic [here](https://github.com/paritytech/substrate/blob/6f72780b87e41a025722284a52a7d85f41958a82/client/executor/wasmtime/src/instance_wrapper.rs#L202), that produced a plain string error of type `Other`.

For now, I'm reverting to the previous behavior: to treat any error reported by the runtime as "not supported". That may cause some minor regression, but I want to believe it's not too critical, we'll see during the burn-in.

I believe that should be properly handled at the lower level, with the Substrate executor returning [`MethodNotFound`](https://github.com/paritytech/substrate/blob/6f72780b87e41a025722284a52a7d85f41958a82/client/executor/common/src/error.rs#L36) and then mapping it into [`RuntimeApiError::NotSupported`](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/subsystem-types/src/errors.rs#L38), but that's out of the scope of this PR.

Filed paritytech/substrate#13344 on that (I'm lucky on beautiful issue and PR numbers).

Reply:
> The behavior pointed out by @tdimitrov only applies when the runtime does not support requesting the API version. When a real runtime method is not supported, another error is returned: 

Can you elaborate on this? The purpose of the version check in the runtime api subsystem is to avoid such cases. We see `Exported method ParachainHost_session_executor_params is not found` so the runtime is old? How come the version check succeeded in such case?

Reply:
@tdimitrov that happens in a staging environment. For example, the current master of Rococo runtime is v3, and my PR is v3 also, but one of them has the method, and another doesn't. Or, if you view it from another angle, yesterday's v3 doesn't have the method, and today's v3 does. In that case, we can only rely on the error reported by the runtime, but right now, that reporting logic is wrong (I believe).

Nothing is wrong with version checking itself in a production environment. If we release v3 with the method and then check versions before calling runtime, everything would work as expected, so no worries here. But that runtime's inability to report a method's absence makes testing really tricky.

Reply:
But the purpose of the versioning is exactly to know what's where. Why there is v3 with the method and another v3 without the method?

Reply:
I'm not sure, probably I'm misunderstanding something, and I'm just confusing people with that misunderstanding. How I see it: version 3 of parachain host API is neither released nor stabilized and is undergoing active development, so we do not yet know a comprehensive list of methods that will be there when it's released. So before the release happens, API versioning is not a reliable source of truth when answering the question, "is the method already there". What am I missing?

Reply:
Until v3 is stabilized we can do whatever we want, but at the moment we deploy it somewhere (even on Versi), for this context v3 is v3. And if you add/remove methods from it - you can expect it to work correctly.

You can bump your method to v4 if you want to test a weird runtime upgrade from v3. But the case you are describing is not supposed to work even in staging.

Reply:
> But the case you are describing is not supposed to work even in staging.

Now I see, but I don't see why. I mean, I see no problem with reporting that from the substrate executor and further using it in client code. Not in the scope of this task, of course, but generally, wouldn't it be useful, at least for testing purposes?

Reply:
The point of checking the version is to be able to gracefully handle an older version. If the function is missing although the version matches, then is a serious error and we would like test networks to blow up ... so we make sure this never hits production, as it would be a bug.